### PR TITLE
feat: point cloud API for styling

### PR DESCRIPTION
**Pull request by haakonflatval-cognite**
This PR partially implements the planned API that were fleshed out in a meeting a few weeks ago: https://cognitedata.atlassian.net/wiki/spaces/REV/pages/3516989550/2022.05.09+-+Point+cloud+styling+API+discussion
The points that are not implemented are:
- Accessor to get the bounding box associated with an annotation
- Downprioritized styles (renderInFront, renderGhosted, outlineColor)
- Picking (worked on by @Savokr )

Otherwise, referring to this list for features that will be implemented to finalize feature branch for merge: https://cognitedata.atlassian.net/browse/REV-374

Edit: I've been a bit hasty on getting this out.. As per our new PR template rules I'm still missing a couple of things:
- [x] More unit tests
- [x] Visual tests
- [x] Developer documentation (`pointcloud` package Readme)
- [x] Documentation
- [ ] ~~Proper documentation example~~
- [x] Check that documentation works

Documentation: I have written some documentation, but have not been able to annotate our example point cloud model with objects due to cumbersome authentication. Thus, the examples are meaningless. Will be fixed later on, when `publicdata` becomes fully OIDC compatible.

### CODE REVIEW

File: examples/src/pages/e2e/pointcloud/ColorStyling.tsx
Comment (larsmoa) [inline examples/src/pages/e2e/pointcloud/ColorStyling.tsx:19]: Suggest you rename this to something a bit more descriptive

File: examples/src/pages/e2e/pointcloud/ColorStyling.tsx
Comment (larsmoa) [inline examples/src/pages/e2e/pointcloud/ColorStyling.tsx:15]: Is Playground SDK still necessary?

File: viewer/packages/api/src/public/createRevealManager.ts
Comment (larsmoa) [inline viewer/packages/api/src/public/createRevealManager.ts:45]: Get rid of playground if SDK is ready

File: viewer/packages/pointclouds/README.md
Comment (larsmoa) [inline viewer/packages/pointclouds/README.md:1]: Love this!

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudFactory.ts:61]: For the future, avoid `autoPaginToArray` and iterate over results as they arrive. I think there is a convenience function for that

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudFactory.ts:72]: Can this be done more type safe? Using non-playground SDK?

If SDK doesn't have the correct types, add a task for that and make sure we have it on the TODO-list.

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudFactory.ts:82]:
```suggestion
      annotations: new Array<PointCloudObjectAnnotation>(),
```

File: viewer/packages/pointclouds/src/PointCloudFactory.test.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudFactory.test.ts:14]: Anything else we could test easily?

File: viewer/packages/pointclouds/src/PointCloudNode.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudNode.ts:182]: Thats a lot of . 

Add convinience accessors to `this._potreeNode`.

File: viewer/packages/pointclouds/src/PointCloudNode.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudNode.ts:100]: This is a confusing name - `set` should only be used when there only can be one, i.e. when calling set will override the effect of previous invocations. Use `assignObjectStyle` (which you have below, so I'm not sure why there's also a setter).

File: viewer/packages/pointclouds/src/PointCloudNode.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudNode.ts:16]: I think this class is uncessary and that we should replace use with two arguments where necessary (i.e. two arguments for assigning styles, one for unassigning).

File: viewer/packages/pointclouds/src/PotreeNodeWrapper.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PotreeNodeWrapper.ts:91]: Why can't it just return `this._annotations` directly? `map()` will allocate. Seems we can get by with only `stylableObjects` below.

File: viewer/packages/pointclouds/src/PotreeNodeWrapper.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PotreeNodeWrapper.ts:101]: There is no `pickWindowSize` anymore. Refer to @Savokr for new approach to point cloud picking - make sure we dont end up duplicating a lot of code here

File: viewer/packages/pointclouds/src/PotreeNodeWrapper.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PotreeNodeWrapper.ts:104]: See comments above recarding naming of this function

File: viewer/packages/pointclouds/src/annotationTypes.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/annotationTypes.ts:37]: I think the number of types can be reduced here

File: viewer/packages/pointclouds/src/annotationTypes.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/annotationTypes.ts:53]: Can this be a proper class, e.g. `PointCloudObjectProvider` or something, with relevant functionality?

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/ObjectAppearanceTexture.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/potree-three-loader/rendering/ObjectAppearanceTexture.ts:12]: Looks unit testable ;)

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/constants.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/potree-three-loader/rendering/constants.ts:18]: Hmm, why is this constant? Doesn't this enforce a limit on 65K objects?

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/shaders/pointcloud.vert
Comment (larsmoa) [inline viewer/packages/pointclouds/src/potree-three-loader/rendering/shaders/pointcloud.vert:606]: Could there be some shared logic with styled CAD objects here? Should be identical, except that the objectIds differ.

File: viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts:5]: Mismatching classname and filename. I suggest we stick with `PointCloudObjectCollection`

File: viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts:10]: See if you can align naming here with naming we have for the corresponding class for CAD

File: viewer/packages/pointclouds/src/styling/PointCloudAppearance.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/PointCloudAppearance.ts:15]:
```suggestion
export function applyDefaultsToPointCloudAppearance(appearance: Partial<PointCloudAppearance>): PointCloudAppearance {
```

File: viewer/packages/pointclouds/src/styling/annotationsToObjects.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/annotationsToObjects.ts:14]: bvs? rename argument

**Review by larsmoa — CHANGES_REQUESTED** [review]

File: examples/src/pages/e2e/pointcloud/ColorStyling.tsx
Comment (haakonflatval-cognite) [reply examples/src/pages/e2e/pointcloud/ColorStyling.tsx:15]: Yes, will fix in a later PR

Comment (haakonflatval-cognite) [reply examples/src/pages/e2e/pointcloud/ColorStyling.tsx]: On second thought, I just pushed a commit where it is removed

File: viewer/packages/pointclouds/README.md
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/README.md:1]: :blush:

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts:61]: Okay - but what would be the benefit of doing that?

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts:72]: The types are not ready in the SDK, @Savokr has taken on the task of adding them there, but not sure when it's ready. Changing these types to the SDK ones are tracked as a bullet point here: https://cognitedata.atlassian.net/browse/REV-374

File: viewer/packages/pointclouds/src/PointCloudNode.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudNode.ts:16]: I disagree - we have the same setup for CAD models, except we haven't bothered giving the type a name there. We need to store what object collection has received what style, and this seems to be the most convenient way of doing so.

File: viewer/packages/pointclouds/src/PotreeNodeWrapper.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PotreeNodeWrapper.ts:91]: Because this method is the one being used to expose the annotations to the user. I have made the choice to only expose the annotation ID and the asset ID, the objects in the `_annotations` list also contain e.g. `objectId` and the geometry information, which I have deemed irrelevant to the user. Do you think we should expose more (apart from a bounding box, which will be a future addition)? On another note, the `_annotations` objects` type is a superset of `AnnotationMetadata`, so we could just return the `_annotations` list anyway, but it would contain more data than just what the type insinuates.

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/constants.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/potree-three-loader/rendering/constants.ts:18]: Yes, this enforces a limit. I can make it higher, but we should probably agree on a reasonable limit. Not sure how easy it would be to make it dynamic.

File: viewer/packages/pointclouds/src/annotationTypes.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/annotationTypes.ts:37]: Removed the geometry types

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/ObjectAppearanceTexture.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/potree-three-loader/rendering/ObjectAppearanceTexture.ts:12]: Added unit tests

File: viewer/packages/pointclouds/src/potree-three-loader/rendering/shaders/pointcloud.vert
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/potree-three-loader/rendering/shaders/pointcloud.vert:606]: Yes, it is similar, but I don't see how we can share logic between the two packages without having to resort to extremely long and brittle `glslify` import strings. Maybe we could refactor to gather shaders in the same spot, but sounds like a job for another PR.

Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/potree-three-loader/rendering/shaders/pointcloud.vert:606]: Tracking it as an issue here: https://cognitedata.atlassian.net/browse/REV-433

File: viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts:10]: Renamed to `AnnotationIdPointCloudObjectCollection`

File: viewer/packages/pointclouds/src/styling/StyledPointCloudObjectCollection.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/styling/StyledPointCloudObjectCollection.ts:9]: Added some documentation

File: viewer/packages/pointclouds/src/CognitePointCloudModel.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/CognitePointCloudModel.ts:246]: This API differs a bit from corresponding CAD API. I think they should be the same, i.e. either `PointCloudAppearance` members becoming optional or `NodeAppearance` members becoming required and use Partial there too (make sure API docs works well if we use Partial)

File: viewer/packages/pointclouds/src/annotationTypes.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/annotationTypes.ts:14]: I think this can be renamed to something more specific, e.g. `PointCloudObjectMapping` or something.

Comment (larsmoa) [reply viewer/packages/pointclouds/src/annotationTypes.ts:14]: Or maybe `PointCloudObjectMetadata`?

File: viewer/packages/pointclouds/src/styling/PointCloudObjectCollection.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/PointCloudObjectCollection.ts:16]: Add TSDOC

**Review by larsmoa — CHANGES_REQUESTED** [review]

File: viewer/package.json
Comment (haakonflatval-cognite) [reply viewer/package.json]: It's not, restricted to 7.7.0 or higher.

File: viewer/packages/pointclouds/src/CognitePointCloudModel.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/CognitePointCloudModel.ts:246]: Made it non-partial, but introduced a `CompletePointCloudObjectAppearance`- type that I use internally instead.

File: viewer/packages/pointclouds/src/annotationTypes.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/annotationTypes.ts:14]: `PointCloudObjectMetadata` it is

File: viewer/packages/pointclouds/src/styling/PointCloudObjectCollection.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/styling/PointCloudObjectCollection.ts:19]: Added some here and in some more of the public-facing part of the code of `CognitePointCloudModel`

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (larsmoa) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts:57]: `autoPagingToArray` will need to allocate an array to hold all elements. Also, it will not allow processing elements in batches as they arrive, postponing all processing until all elements are downloaded (which could take a while with O(100.000) elements).

File: documentation/docs/examples/pointcloud-styling.mdx
Comment (larsmoa) [inline documentation/docs/examples/pointcloud-styling.mdx:59]:
```suggestion
## Resetting styled objects
```

File: documentation/docs/examples/pointcloud-styling.mdx
Comment (larsmoa) [inline documentation/docs/examples/pointcloud-styling.mdx:61]:
```suggestion
To revert an assignment of a style to an object collection, you may use `model.unassignStyledObjectCollection`, using the `PointCloudObjectCollection` as argument. To reset all styled objects use `model.removeAllStyledCollections`
```

File: examples/src/utils/PointCloudObjectStylingUI.ts
Comment (larsmoa) [inline examples/src/utils/PointCloudObjectStylingUI.ts:29]:
```suggestion
    uiFolder.add(actions, 'reset').name('Reset all styled objects');
```

File: viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts
Comment (larsmoa) [inline viewer/packages/pointclouds/src/styling/AnnotationListPointCloudObjectCollection.ts:18]:
```suggestion
  constructor(ids: Iterable<number>) {
    super();
    this._annotations = new Set<number>(ids);
  }
```
should work and be more generic

**Review by larsmoa — CHANGES_REQUESTED** [review]

File: viewer/packages/pointclouds/src/PointCloudNode.ts
Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudNode.ts:182]: Ah, sorry - I did not intend for it to be exported. Will fix

File: viewer/packages/pointclouds/src/PointCloudFactory.ts
Comment (andersfylling) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts:28]: The PR is done coding wise, just waiting for review. Note that it's going into a test phase after this, so once that is merged, types won't automatically be generated for everyone. It has to be enabled manually for a specific service and then changes have to be manually approved.

Note that the codegen PR only generates request/response types defined in a provided open api spec. By default cognite packages use https://docs.cognite.com/api/v1/, but I don't see pointcloud there?

Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts:28]: @andersfylling I put this comment (which I incidentally deleted before I saw your response)  in the wrong spot - see the conversation directly above for response

Comment (haakonflatval-cognite) [reply viewer/packages/pointclouds/src/PointCloudFactory.ts]: I have created separation between `CdfPointCloudFactory` and `LocalPointCloudFactory` now.

**Pull request merged**
